There is a test failure with:
```
Caused by: java.lang.NoClassDefFoundError: org/jboss/resteasy/core/ResteasyContext
	at org.jboss.resteasy.context.ResteasyContextProvider.currentContext(ResteasyContextProvider.java:18)
	at io.smallrye.context.impl.CapturedContextState.<init>(CapturedContextState.java:21)
	at io.smallrye.context.SmallRyeContextManager.captureContext(SmallRyeContextManager.java:64)
	at io.smallrye.context.SmallRyeThreadContext.currentContextExecutor(SmallRyeThreadContext.java:208)
	at io.smallrye.mutiny.context.ContextPropagationUniInterceptor.onUniCreation(ContextPropagationUniInterceptor.java:47)
	at io.smallrye.mutiny.infrastructure.Infrastructure.onUniCreation(Infrastructure.java:70)
	at io.smallrye.mutiny.groups.UniOnItem.apply(UniOnItem.java:50)
	at io.quarkus.vertx.EventBusCodecTest.testWithUserCodecNonLocal(EventBusCodecTest.java:59)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at io.quarkus.test.QuarkusUnitTest.runExtensionMethod(QuarkusUnitTest.java:276)
```

--- a/bom/runtime/pom.xml
+++ b/bom/runtime/pom.xml
@@ -2506,6 +2506,11 @@
                 <artifactId>smallrye-reactive-converter-api</artifactId>
                 <version>${smallrye-converter-api.version}</version>
             </dependency>
+            <dependency>
+                <groupId>io.smallrye.reactive</groupId>
+                <artifactId>smallrye-reactive-converter-rxjava2</artifactId>
+                <version>${smallrye-converter-api.version}</version>
+            </dependency>
 
             <dependency>
                 <groupId>org.jetbrains.kotlin</groupId>

--- a/bom/runtime/pom.xml
+++ b/bom/runtime/pom.xml
@@ -996,6 +996,11 @@
                 <version>${mutiny.version}</version>
             </dependency>
             <dependency>
+                <groupId>io.smallrye.reactive</groupId>
+                <artifactId>mutiny-context-propagation</artifactId>
+                <version>${mutiny.version}</version>
+            </dependency>
+            <dependency>
                 <groupId>io.quarkus</groupId>
                 <artifactId>quarkus-mutiny</artifactId>
                 <version>${project.version}</version>

--- a/extensions/resteasy-mutiny/deployment/src/test/java/io/quarkus/resteasy/mutiny/test/MutinyResource.java
+++ b/extensions/resteasy-mutiny/deployment/src/test/java/io/quarkus/resteasy/mutiny/test/MutinyResource.java
@@ -3,8 +3,10 @@ package io.quarkus.resteasy.mutiny.test;
 import javax.ws.rs.GET;
 import javax.ws.rs.Path;
 import javax.ws.rs.Produces;
+import javax.ws.rs.WebApplicationException;
 import javax.ws.rs.core.Context;
 import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
 
 import org.jboss.resteasy.annotations.Stream;
 
@@ -38,5 +40,22 @@ public class MutinyResource {
     @GET
     public Uni<Integer> injectionAsync(@Async @Context Integer value) {
         return Uni.createFrom().item(value);
+    }
+
+    @Path("web-failure")
+    @GET
+    public Uni<String> failing() {
+        return Uni.createFrom().item("not ok")
+                .onItem().failWith(s -> new WebApplicationException(
+                        Response.status(Response.Status.SERVICE_UNAVAILABLE).entity(s).build()));
+    }
+
+    @Path("app-failure")
+    @GET
+    public Uni<String> failingBecauseOfApplicationCode() {
+        return Uni.createFrom().item("not ok")
+                .onItem().apply(s -> {
+                    throw new IllegalStateException("BOOM!");
+                });
     }
 }

--- a/extensions/resteasy-mutiny/deployment/src/test/java/io/quarkus/resteasy/mutiny/test/RestEasyMutinyTest.java
+++ b/extensions/resteasy-mutiny/deployment/src/test/java/io/quarkus/resteasy/mutiny/test/RestEasyMutinyTest.java
@@ -6,6 +6,7 @@ import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicReference;
 
 import javax.ws.rs.client.ClientBuilder;
+import javax.ws.rs.core.Response;
 
 import org.jboss.resteasy.client.jaxrs.ResteasyClient;
 import org.jboss.resteasy.client.jaxrs.ResteasyClientBuilder;
@@ -58,6 +59,18 @@ public class RestEasyMutinyTest {
 
         data = client.target(url.toExternalForm() + "/injection-async").request().get(Integer.class);
         Assertions.assertEquals((Integer) 42, data);
+    }
+
+    @Test
+    public void testFailingWithWebApplicationException() {
+        Response response = client.target(url.toExternalForm() + "/web-failure").request().get();
+        Assertions.assertEquals(response.getStatus(), Response.Status.SERVICE_UNAVAILABLE.getStatusCode());
+    }
+
+    @Test
+    public void testFailingWithApplicationFailure() {
+        Response response = client.target(url.toExternalForm() + "/app-failure").request().get();
+        Assertions.assertEquals(response.getStatus(), 500);
     }
 
 }

--- a/extensions/smallrye-context-propagation/deployment/pom.xml
+++ b/extensions/smallrye-context-propagation/deployment/pom.xml
@@ -56,7 +56,12 @@
         <dependency>
             <groupId>io.smallrye.reactive</groupId>
             <artifactId>smallrye-reactive-converter-rxjava2</artifactId>
-            <version>1.0.4</version>
+            <scope>test</scope>
+        </dependency>
+        <!-- For Mutiny tests -->
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-resteasy-mutiny-deployment</artifactId>
             <scope>test</scope>
         </dependency>
         <!-- For DB -->

--- a/extensions/smallrye-context-propagation/deployment/src/test/java/io/quarkus/context/test/mutiny/MutinyContextEndpoint.java
+++ b/extensions/smallrye-context-propagation/deployment/src/test/java/io/quarkus/context/test/mutiny/MutinyContextEndpoint.java
@@ -1,0 +1,461 @@
+package io.quarkus.context.test.mutiny;
+
+import java.net.MalformedURLException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+import javax.servlet.http.HttpServletRequest;
+import javax.transaction.Status;
+import javax.transaction.SystemException;
+import javax.transaction.Transaction;
+import javax.transaction.Transactional;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.UriInfo;
+
+import org.eclipse.microprofile.context.ManagedExecutor;
+import org.eclipse.microprofile.context.ThreadContext;
+import org.jboss.resteasy.annotations.Stream;
+import org.jboss.resteasy.annotations.Stream.MODE;
+import org.junit.jupiter.api.Assertions;
+import org.reactivestreams.Publisher;
+import org.wildfly.common.Assert;
+
+import io.quarkus.arc.Arc;
+import io.quarkus.context.test.RequestBean;
+import io.quarkus.hibernate.orm.panache.Panache;
+import io.smallrye.mutiny.Multi;
+import io.smallrye.mutiny.Uni;
+import io.smallrye.mutiny.infrastructure.Infrastructure;
+
+@Path("/mutiny-context")
+@Produces(MediaType.TEXT_PLAIN)
+public class MutinyContextEndpoint {
+
+    @Inject
+    RequestBean doNotRemoveMe;
+    @Inject
+    ManagedExecutor all;
+    @Inject
+    ThreadContext allTc;
+    @Inject
+    HttpServletRequest servletRequest;
+
+    ExecutorService executor = Executors.newSingleThreadExecutor();
+
+    @PreDestroy
+    public void shutdown() {
+        executor.shutdown();
+    }
+
+    @GET
+    @Path("/resteasy-uni-cs")
+    public Uni<String> resteasyContextPropagationWithUniCreatedFromCSWithManagedExecutor(@Context UriInfo uriInfo) {
+        CompletableFuture<String> ret = all.completedFuture("OK");
+        return Uni.createFrom().completionStage(ret)
+                .emitOn(Infrastructure.getDefaultExecutor())
+                .onItem().apply(s -> {
+                    Assertions.assertNotNull(uriInfo.getAbsolutePath());
+                    try {
+                        Assertions.assertTrue(
+                                uriInfo.getAbsolutePath().toURL().toExternalForm().contains("/resteasy-uni-cs"));
+                    } catch (MalformedURLException e) {
+                        throw new AssertionError(e);
+                    }
+                    return s;
+                });
+    }
+
+    @GET
+    @Path("/resteasy-uni")
+    public Uni<String> resteasyContextPropagation(@Context UriInfo uriInfo) {
+        return Uni.createFrom().item("OK")
+                .emitOn(Infrastructure.getDefaultExecutor())
+                .onItem().apply(s -> {
+                    Assertions.assertNotNull(uriInfo.getAbsolutePath());
+                    try {
+                        Assertions.assertTrue(
+                                uriInfo.getAbsolutePath().toURL().toExternalForm().contains("/resteasy-uni"));
+                    } catch (MalformedURLException e) {
+                        throw new AssertionError(e);
+                    }
+                    return s;
+                });
+    }
+
+    @GET
+    @Path("/resteasy-tc-uni-cs")
+    public Uni<String> resteasyThreadContextWithCS(@Context UriInfo uriInfo) {
+        CompletableFuture<String> ret = allTc.withContextCapture(CompletableFuture.completedFuture("OK"));
+        return Uni.createFrom().completionStage(ret)
+                .emitOn(executor)
+                .onItem().apply(s -> {
+                    uriInfo.getAbsolutePath();
+                    return s;
+                });
+    }
+
+    @GET
+    @Path("/servlet-uni")
+    public Uni<String> servletContextPropagation(@Context UriInfo uriInfo) {
+        return Uni.createFrom().item("OK")
+                .emitOn(Infrastructure.getDefaultExecutor())
+                .map(text -> {
+                    Assertions.assertNotNull(servletRequest.getContentType());
+                    return text;
+                })
+                .onFailure().invoke(t -> System.out.println("Got failure " + t.getMessage()));
+    }
+
+    @GET
+    @Path("/servlet-uni-cs")
+    public Uni<String> servletContextPropagationWithUniCreatedFromCSWithManagedExecutor(@Context UriInfo uriInfo) {
+        CompletableFuture<String> ret = all.completedFuture("OK");
+        return Uni.createFrom().completionStage(() -> ret)
+                .emitOn(Infrastructure.getDefaultExecutor())
+                .map(text -> {
+                    Assertions.assertNotNull(servletRequest.getContentType());
+                    return text;
+                });
+    }
+
+    @GET
+    @Path("/servlet-tc-uni-cs")
+    public Uni<String> servletThreadContext(@Context UriInfo uriInfo) {
+        CompletableFuture<String> ret = allTc.withContextCapture(CompletableFuture.completedFuture("OK"));
+        return Uni.createFrom().completionStage(() -> ret)
+                .emitOn(executor)
+                .map(text -> {
+                    Assertions.assertNotNull(servletRequest.getContentType());
+                    return text;
+                });
+    }
+
+    @GET
+    @Path("/arc-uni")
+    public Uni<String> arcContextPropagation() {
+        Assert.assertTrue(Arc.container().instance(RequestBean.class).isAvailable());
+        RequestBean instance = Arc.container().instance(RequestBean.class).get();
+        String previousValue = instance.callMe();
+        return Uni.createFrom().item("OK")
+                .emitOn(Infrastructure.getDefaultExecutor())
+                .map(text -> {
+                    RequestBean instance2 = Arc.container().instance(RequestBean.class).get();
+                    Assertions.assertEquals(previousValue, instance2.callMe());
+                    return text;
+                });
+    }
+
+    @GET
+    @Path("/arc-uni-cs")
+    public Uni<String> arcContextPropagationWithUniCreatedFromCSWithManagedExecutor() {
+        Assert.assertTrue(Arc.container().instance(RequestBean.class).isAvailable());
+        RequestBean instance = Arc.container().instance(RequestBean.class).get();
+        String previousValue = instance.callMe();
+        CompletableFuture<String> ret = all.completedFuture("OK");
+        return Uni.createFrom().completionStage(() -> ret)
+                .emitOn(Infrastructure.getDefaultExecutor())
+                .map(text -> {
+                    RequestBean instance2 = Arc.container().instance(RequestBean.class).get();
+                    Assertions.assertEquals(previousValue, instance2.callMe());
+                    return text;
+                });
+    }
+
+    @GET
+    @Path("/arc-tc-uni")
+    public Uni<String> arcContextPropagationWithThreadContext() {
+        ExecutorService executor = Executors.newSingleThreadExecutor();
+
+        Assert.assertTrue(Arc.container().instance(RequestBean.class).isAvailable());
+        RequestBean instance = Arc.container().instance(RequestBean.class).get();
+        String previousValue = instance.callMe();
+        CompletableFuture<String> ret = allTc.withContextCapture(CompletableFuture.completedFuture("OK"));
+        return Uni.createFrom().completionStage(() -> ret)
+                .emitOn(executor)
+                .map(text -> {
+                    RequestBean instance2 = Arc.container().instance(RequestBean.class).get();
+                    Assertions.assertEquals(previousValue, instance2.callMe());
+                    return text;
+                });
+    }
+
+    @Inject
+    MutinyTransactionalBean txBean;
+
+    @Transactional
+    @GET
+    @Path("/transaction-uni")
+    public Uni<String> contextPropagationWithTxAndUni() throws SystemException {
+        SomeEntity.deleteAll();
+        Uni<String> ret = Uni.createFrom().item("OK");
+        SomeEntity entity = new SomeEntity();
+        entity.name = "Stef";
+        entity.persist();
+        Transaction t1 = Panache.getTransactionManager().getTransaction();
+        Assertions.assertNotNull(t1);
+
+        return ret
+                .emitOn(executor)
+                .map(text -> {
+                    Assertions.assertEquals(1, SomeEntity.count());
+                    Transaction t2;
+                    try {
+                        t2 = Panache.getTransactionManager().getTransaction();
+                    } catch (SystemException e) {
+                        throw new RuntimeException(e);
+                    }
+                    Assertions.assertEquals(t1, t2);
+                    return text;
+                });
+    }
+
+    @Transactional
+    @GET
+    @Path("/transaction-uni-cs")
+    public Uni<String> contextPropagationWithTxAndUniCreatedFromCS() throws SystemException {
+        Uni<String> ret = Uni.createFrom().completionStage(all.completedFuture("OK"));
+        SomeOtherEntity entity = new SomeOtherEntity();
+        entity.name = "Stef";
+        entity.persist();
+        Transaction t1 = Panache.getTransactionManager().getTransaction();
+        Assertions.assertNotNull(t1);
+
+        return ret
+                .emitOn(executor)
+                .map(text -> {
+                    Assertions.assertEquals(1, SomeOtherEntity.count());
+                    Transaction t2;
+                    try {
+                        t2 = Panache.getTransactionManager().getTransaction();
+                    } catch (SystemException e) {
+                        throw new RuntimeException(e);
+                    }
+                    Assertions.assertEquals(t1, t2);
+                    return text;
+                });
+    }
+
+    @Transactional
+    @GET
+    @Path("/transaction-tc-uni")
+    public Uni<String> transactionPropagationWithThreadContextAndUniCreatedFromCS() throws SystemException {
+        CompletableFuture<String> ret = allTc.withContextCapture(CompletableFuture.completedFuture("OK"));
+        SomeEntity entity = new SomeEntity();
+        entity.name = "Stef";
+        entity.persist();
+        Transaction t1 = Panache.getTransactionManager().getTransaction();
+        Assertions.assertNotNull(t1);
+
+        return Uni.createFrom().completionStage(ret)
+                .emitOn(executor)
+                .map(text -> {
+                    Assertions.assertEquals(1, SomeEntity.count());
+                    Transaction t2;
+                    try {
+                        t2 = Panache.getTransactionManager().getTransaction();
+                    } catch (SystemException e) {
+                        throw new RuntimeException(e);
+                    }
+                    Assertions.assertEquals(t1, t2);
+                    return text;
+                });
+    }
+
+    @Transactional
+    @GET
+    @Path("/transaction2-uni")
+    public Uni<String> transactionTest2() {
+        Uni<String> ret = Uni.createFrom().item("OK");
+
+        // check that the first transaction was committed
+        Assertions.assertEquals(1, SomeEntity.count());
+        // now delete our entity, but throw an exception to rollback
+        Assertions.assertEquals(1, SomeEntity.deleteAll());
+
+        return ret
+                .emitOn(executor)
+                .onItem().failWith(s -> new WebApplicationException(Response.status(Response.Status.CONFLICT).build()));
+    }
+
+    @Transactional
+    @GET
+    @Path("/transaction2-uni-cs")
+    public Uni<String> transactionTest2WithUniCreatedFromCS() {
+        Uni<String> ret = Uni.createFrom().completionStage(all.completedFuture("OK"));
+
+        // check that the first transaction was committed
+        Assertions.assertEquals(1, SomeOtherEntity.count());
+        // now delete our entity, but throw an exception to rollback
+        Assertions.assertEquals(1, SomeOtherEntity.deleteAll());
+
+        return ret
+                .emitOn(executor)
+                .onItem().failWith(s -> new WebApplicationException(Response.status(Response.Status.CONFLICT).build()));
+    }
+
+    @Transactional
+    @GET
+    @Path("/transaction3-uni")
+    public Uni<String> transactionTest3() {
+        Uni<String> ret = Uni.createFrom()
+                .failure(new WebApplicationException(Response.status(Response.Status.CONFLICT).build()));
+
+        // check that the second transaction was not committed
+        Assertions.assertEquals(1, SomeEntity.count());
+        // now delete our entity, but throw an exception to rollback
+        Assertions.assertEquals(1, SomeEntity.deleteAll());
+
+        return ret;
+    }
+
+    @Transactional
+    @GET
+    @Path("/transaction3-uni-cs")
+    public Uni<String> transactionTest3WithUniCreatedFromCS() {
+        CompletableFuture<String> future = all
+                .failedFuture(new WebApplicationException(Response.status(Response.Status.CONFLICT).build()));
+        Uni<String> ret = Uni.createFrom().completionStage(future);
+
+        // check that the second transaction was not committed
+        Assertions.assertEquals(1, SomeOtherEntity.count());
+        // now delete our entity, but throw an exception to rollback
+        Assertions.assertEquals(1, SomeOtherEntity.deleteAll());
+
+        return ret;
+    }
+
+    @Transactional
+    @GET
+    @Path("/transaction4")
+    public String transactionTest4() {
+        // check that the third transaction was not committed
+        Assertions.assertEquals(1, SomeEntity.count());
+        // now delete our entity
+        Assertions.assertEquals(1, SomeEntity.deleteAll());
+
+        return "OK";
+    }
+
+    @Transactional
+    @GET
+    @Path("/transaction4-cs")
+    public String transactionTest4CS() {
+        // check that the third transaction was not committed
+        Assertions.assertEquals(1, SomeOtherEntity.count());
+        // now delete our entity
+        Assertions.assertEquals(1, SomeOtherEntity.deleteAll());
+
+        return "OK";
+    }
+
+    @Transactional
+    @GET
+    @Path("/transaction-new-sync")
+    public Uni<String> newTransactionPropagationSynchronous() throws SystemException {
+        Uni<String> ret = Uni.createFrom().item("OK");
+        Transaction t1 = Panache.getTransactionManager().getTransaction();
+        Assertions.assertNotNull(t1);
+
+        txBean.doInTx();
+
+        // We should see the transaction already committed even if we're async
+        Assertions.assertEquals(1, Person.deleteAll());
+        return ret;
+    }
+
+    @Transactional
+    @GET
+    @Path("/transaction-new-uni")
+    public Uni<String> newTransactionPropagationWithUni() throws SystemException {
+        Person entity = new Person();
+        entity.name = "Stef";
+        entity.persist();
+        Transaction t1 = Panache.getTransactionManager().getTransaction();
+        Assertions.assertNotNull(t1);
+        // our entity
+        Assertions.assertEquals(1, Person.count());
+
+        return txBean.doInTxUni()
+                // this makes sure we get executed in another scheduler
+                .emitOn(executor)
+                .map(text -> {
+                    // make sure we don't see the other transaction's entity
+                    Transaction t2;
+                    try {
+                        t2 = Panache.getTransactionManager().getTransaction();
+                    } catch (SystemException e) {
+                        throw new RuntimeException(e);
+                    }
+                    Assertions.assertEquals(t1, t2);
+                    try {
+                        Assertions.assertEquals(Status.STATUS_ACTIVE, t2.getStatus());
+                    } catch (SystemException e) {
+                        throw new AssertionError(e);
+                    }
+                    return text;
+                });
+    }
+
+    @Transactional
+    @GET
+    @Path("/transaction-uni-2")
+    public Uni<String> transactionPropagationWithUni() {
+        Uni<String> ret = Uni.createFrom().item("OK");
+        // now delete both entities
+        Assertions.assertEquals(2, Person.deleteAll());
+        return ret;
+    }
+
+    @Transactional
+    @GET
+    @Path("/transaction-multi")
+    @Stream(value = MODE.RAW)
+    public Multi<String> transactionPropagationWithMulti() throws SystemException {
+        Person entity = new Person();
+        entity.name = "Stef";
+        entity.persist();
+        Transaction t1 = Panache.getTransactionManager().getTransaction();
+        Assertions.assertNotNull(t1);
+        // our entity
+        Assertions.assertEquals(1, Person.count());
+
+        return txBean.doInTxMulti()
+                // this makes sure we get executed in another scheduler
+                .emitOn(Infrastructure.getDefaultExecutor())
+                .map(text -> {
+                    // make sure we don't see the other transaction's entity
+                    Transaction t2;
+                    try {
+                        t2 = Panache.getTransactionManager().getTransaction();
+                    } catch (SystemException e) {
+                        throw new RuntimeException(e);
+                    }
+                    Assertions.assertEquals(t1, t2);
+                    try {
+                        Assertions.assertEquals(Status.STATUS_ACTIVE, t2.getStatus());
+                    } catch (SystemException e) {
+                        throw new AssertionError(e);
+                    }
+                    return text;
+                });
+    }
+
+    @Transactional
+    @GET
+    @Path("/transaction-multi-2")
+    public Publisher<String> transactionPropagationWithMulti2() {
+        Multi<String> ret = Multi.createFrom().item("OK");
+        // now delete both entities
+        Assertions.assertEquals(2, Person.deleteAll());
+        return ret;
+    }
+}

--- a/extensions/smallrye-context-propagation/deployment/src/test/java/io/quarkus/context/test/mutiny/MutinyContextPropagationTest.java
+++ b/extensions/smallrye-context-propagation/deployment/src/test/java/io/quarkus/context/test/mutiny/MutinyContextPropagationTest.java
@@ -1,0 +1,156 @@
+package io.quarkus.context.test.mutiny;
+
+import static org.hamcrest.Matchers.equalTo;
+
+import javax.ws.rs.core.Response;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.context.test.CompletionExceptionMapper;
+import io.quarkus.context.test.RequestBean;
+import io.quarkus.context.test.TestResources;
+import io.quarkus.test.QuarkusUnitTest;
+import io.restassured.RestAssured;
+
+/**
+ * Tests Mutiny context propagation of basic contexts (Arc, RestEasy, server, transaction) via either
+ * {@link org.eclipse.microprofile.context.ManagedExecutor} (ME) or
+ * {@link org.eclipse.microprofile.context.ThreadContext} (TC).
+ */
+public class MutinyContextPropagationTest {
+    private static Class[] testClasses = {
+            MutinyContextEndpoint.class, RequestBean.class, SomeEntity.class, SomeOtherEntity.class, Person.class,
+            TestResources.class,
+            CompletionExceptionMapper.class,
+            MutinyTransactionalBean.class
+    };
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
+                    .addClasses(testClasses)
+                    .addAsResource("application.properties"));
+
+    @Test
+    public void testRESTEasyContextPropagation() {
+        RestAssured.when().get("/mutiny-context/resteasy-uni").then()
+                .statusCode(Response.Status.OK.getStatusCode());
+    }
+
+    @Test
+    public void testRESTEasyMEContextPropagationWithUniCreatedFromCS() {
+        RestAssured.when().get("/mutiny-context/resteasy-uni-cs").then()
+                .statusCode(Response.Status.OK.getStatusCode());
+    }
+
+    @Test
+    public void testRESTEasyTCContextPropagationWithUniCreatedFromCS() {
+        RestAssured.when().get("/mutiny-context/resteasy-tc-uni-cs").then()
+                .statusCode(Response.Status.OK.getStatusCode());
+    }
+
+    @Test
+    public void testServletContextPropagation() {
+        RestAssured
+                .given()
+                .header("Content-Type", "acme/acme")
+                .when()
+                .get("/mutiny-context/servlet-uni")
+                .then().statusCode(Response.Status.OK.getStatusCode());
+    }
+
+    @Test
+    public void testServletMEContextPropagationWithUniCreatedFromCS() {
+        RestAssured
+                .given()
+                .header("Content-Type", "acme/acme")
+                .when()
+                .get("/mutiny-context/servlet-uni-cs")
+                .then().statusCode(Response.Status.OK.getStatusCode());
+    }
+
+    @Test
+    public void testServletTCContextPropagationWithUniCreatedFromCS() {
+        RestAssured
+                .given()
+                .header("Content-Type", "acme/acme")
+                .when()
+                .get("/mutiny-context/servlet-tc-uni-cs")
+                .then().statusCode(Response.Status.OK.getStatusCode());
+    }
+
+    @Test
+    public void testArcContextPropagationWithUni() {
+        RestAssured.when().get("/mutiny-context/arc-uni").then()
+                .statusCode(Response.Status.OK.getStatusCode());
+    }
+
+    @Test
+    public void testArcContextPropagationWithUniCreatedFromCSAndME() {
+        RestAssured.when().get("/mutiny-context/arc-uni-cs").then()
+                .statusCode(Response.Status.OK.getStatusCode());
+    }
+
+    @Test
+    public void testArcTCContextPropagation() {
+        RestAssured.when().get("/mutiny-context/arc-tc-uni").then()
+                .statusCode(Response.Status.OK.getStatusCode());
+    }
+
+    @Test
+    public void testTransactionPropagationWithUni() {
+        RestAssured.when().get("/mutiny-context/transaction-uni").then()
+                .statusCode(Response.Status.OK.getStatusCode());
+        RestAssured.when().get("/mutiny-context/transaction2-uni").then()
+                .statusCode(Response.Status.CONFLICT.getStatusCode());
+        RestAssured.when().get("/mutiny-context/transaction3-uni").then()
+                .statusCode(Response.Status.CONFLICT.getStatusCode());
+        RestAssured.when().get("/mutiny-context/transaction4").then()
+                .statusCode(Response.Status.OK.getStatusCode());
+    }
+
+    @Test
+    public void testTransactionPropagationWithUniCreatedFromCS() {
+        RestAssured.when().get("/mutiny-context/transaction-uni-cs").then()
+                .statusCode(Response.Status.OK.getStatusCode());
+        RestAssured.when().get("/mutiny-context/transaction2-uni-cs").then()
+                .statusCode(Response.Status.CONFLICT.getStatusCode());
+        RestAssured.when().get("/mutiny-context/transaction3-uni-cs").then()
+                .statusCode(Response.Status.CONFLICT.getStatusCode());
+        RestAssured.when().get("/mutiny-context/transaction4-cs").then()
+                .statusCode(Response.Status.OK.getStatusCode());
+    }
+
+    @Test
+    public void testTransactionTCContextPropagation() {
+        RestAssured.when().get("/mutiny-context/transaction-tc-uni").then()
+                .statusCode(Response.Status.OK.getStatusCode());
+    }
+
+    @Test
+    public void testTransactionNewContextPropagationSync() {
+        RestAssured.when().get("/mutiny-context/transaction-new-sync").then()
+                .statusCode(Response.Status.OK.getStatusCode());
+    }
+
+    @Test
+    public void testTransactionContextPropagationAsyncUni() {
+        RestAssured.when().get("/mutiny-context/transaction-new-uni").then()
+                .statusCode(Response.Status.OK.getStatusCode())
+                .body(equalTo("OK"));
+        RestAssured.when().get("/mutiny-context/transaction-uni-2").then()
+                .statusCode(Response.Status.OK.getStatusCode());
+    }
+
+    @Test
+    public void testTransactionContextPropagationMulti() {
+        RestAssured.when().get("/mutiny-context/transaction-multi").then()
+                .statusCode(Response.Status.OK.getStatusCode())
+                .body(equalTo("OK"));
+        RestAssured.when().get("/mutiny-context/transaction-multi-2").then()
+                .statusCode(Response.Status.OK.getStatusCode());
+    }
+}

--- a/extensions/smallrye-context-propagation/deployment/src/test/java/io/quarkus/context/test/mutiny/MutinyTransactionalBean.java
+++ b/extensions/smallrye-context-propagation/deployment/src/test/java/io/quarkus/context/test/mutiny/MutinyTransactionalBean.java
@@ -1,0 +1,50 @@
+package io.quarkus.context.test.mutiny;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+import javax.transaction.TransactionManager;
+import javax.transaction.Transactional;
+import javax.transaction.Transactional.TxType;
+
+import org.junit.jupiter.api.Assertions;
+
+import io.smallrye.mutiny.Multi;
+import io.smallrye.mutiny.Uni;
+
+@ApplicationScoped
+public class MutinyTransactionalBean {
+
+    @Inject
+    TransactionManager tm;
+
+    @Transactional(value = TxType.REQUIRES_NEW)
+    public void doInTx() {
+        Assertions.assertEquals(0, Person.count());
+
+        Person entity = new Person();
+        entity.name = "Stef";
+        entity.persist();
+    }
+
+    @Transactional(value = TxType.REQUIRES_NEW)
+    public Uni<String> doInTxUni() {
+        Assertions.assertEquals(0, Person.count());
+
+        Person entity = new Person();
+        entity.name = "Stef";
+        entity.persist();
+
+        return Uni.createFrom().item("OK");
+    }
+
+    @Transactional(value = TxType.REQUIRES_NEW)
+    public Multi<String> doInTxMulti() {
+        Assertions.assertEquals(0, Person.count());
+
+        Person entity = new Person();
+        entity.name = "Stef";
+        entity.persist();
+
+        return Multi.createFrom().items("OK");
+    }
+}

--- a/extensions/smallrye-context-propagation/deployment/src/test/java/io/quarkus/context/test/mutiny/Person.java
+++ b/extensions/smallrye-context-propagation/deployment/src/test/java/io/quarkus/context/test/mutiny/Person.java
@@ -1,0 +1,11 @@
+package io.quarkus.context.test.mutiny;
+
+import javax.persistence.Entity;
+
+import io.quarkus.hibernate.orm.panache.PanacheEntity;
+
+@Entity
+public class Person extends PanacheEntity {
+
+    public String name;
+}

--- a/extensions/smallrye-context-propagation/deployment/src/test/java/io/quarkus/context/test/mutiny/SomeEntity.java
+++ b/extensions/smallrye-context-propagation/deployment/src/test/java/io/quarkus/context/test/mutiny/SomeEntity.java
@@ -1,0 +1,12 @@
+package io.quarkus.context.test.mutiny;
+
+import javax.persistence.Entity;
+
+import io.quarkus.hibernate.orm.panache.PanacheEntity;
+
+@Entity
+public class SomeEntity extends PanacheEntity {
+
+    public String name;
+
+}

--- a/extensions/smallrye-context-propagation/deployment/src/test/java/io/quarkus/context/test/mutiny/SomeOtherEntity.java
+++ b/extensions/smallrye-context-propagation/deployment/src/test/java/io/quarkus/context/test/mutiny/SomeOtherEntity.java
@@ -1,0 +1,12 @@
+package io.quarkus.context.test.mutiny;
+
+import javax.persistence.Entity;
+
+import io.quarkus.hibernate.orm.panache.PanacheEntity;
+
+@Entity
+public class SomeOtherEntity extends PanacheEntity {
+
+    public String name;
+
+}

--- a/extensions/smallrye-context-propagation/runtime/pom.xml
+++ b/extensions/smallrye-context-propagation/runtime/pom.xml
@@ -39,6 +39,10 @@
             </exclusions>
         </dependency>
         <dependency>
+            <groupId>io.smallrye.reactive</groupId>
+            <artifactId>mutiny-context-propagation</artifactId>
+        </dependency>
+        <dependency>
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-core</artifactId>
         </dependency>

